### ISSUE TITLE
DDRec in python

### ISSUE THREAD

**Issue by rete** [AIDASoft/DD4hep#306]
Hi DD4hep devs !

I looked in the available tools in the DD4hep package and I haven't found a python interface for DDRec as it is available for DDG4 (and thus ddsim). 
I have no idea how difficult it is and how much effort I would take to make this interface.
I want to use it for the calibration software, LCCalibration, mainly to to compute the total interaction length in  sub-detectors, in particular in the calorimeters. 
As an ugly work-around, I'm converting the compact file to a Gear file and then parse it with the lxml python module. It currently works, but as we would like to get rid of Gear in the ILD reconstruction chain and also like to remove this (ugly) conversion from the calibration software.

**Comment by decibelcooper** [comment]
I am not one of the devs, but:

The Python written for DDG4 is helper code on top of pyROOT.  You can use any part of DD4Hep or other libraries inside of Python with pyROOT.  It's easiest with dictionaries, but it's also possible without, by telling ROOT to process relevant headers (ROOT.gInterpreter.ProcessLine()).

In any case, if what you are looking for is direct access to the geometry, that is no problem:
```python
import DD4hep
det = DD4hep.Detector.getInstance()
det.fromXML('compact.xml')
```

**Comment by MarkusFrankATcernch** [comment]
@decibelcooper
Absolutely correct. In principle anything which has a ROOT dictionary can be used together with the DDG4 or DD4hep python interfaces. In principle could the entire detector be instantiated using python 
(ie. python detector constructors), but I do not like this.

@gaede would it be reasonable to investigate the dictionary creation for DDRec ?
@rete : You should be aware, that crossing language boundaries is never for free. To steer things this is of course OK, but whenever you intend to do serious number crunching, the price is significant.

**Comment by gaede** [comment]
This would clearly be *nice to have* - as we had already discussed in one of our developers meetings a while ago.
The first obvious step would be to add a dictionary to DDRec. 
@MarkusFrankATcernch can this be done rather easily with your cmake macros ?
The second step, which requires a bit more thought, is how to provide access to the DetectorData user extensions that are really what @rete is interested in.
This could either be done by adding dedicated C++ methods returning the concrete type - or by defining suitable handle classes.

**Comment by gaede** [comment]
I have now implemented dictionary creation for DDRec, addressed the access to the DetectorData objects and added a DDRec.py:  #312.  However, when loading this in python, I do not see any of my classes. So sth. seems to be missing - probably in https://github.com/gaede/DD4hep/blob/630f37ec89624645dee7f848722927a4dc5c8b2d/DDRec/python/DDRec.py.
@MarkusFrankATcernch can you have a quick look at the file - it is a copy from your DDG4.py where I commented out what I believed to not be needed, obviously too much.

**Comment by MarkusFrankATcernch** [comment]
Hi Frank,

use this here. These dictionaries for DDRec work.
https://github.com/MarkusFrankATcernch/DD4hep

**Comment by gaede** [comment]
Markus, looks good - thanks. At least I can see the DDRec classes after 
`import DDRec`. Please merge it - I will then drop #312.

**Comment by MarkusFrankATcernch** [comment]
Frank,
once https://github.com/AIDASoft/DD4hep/pull/325 is merged,
could you please close the issue?

**Comment by gaede** [comment]
There is still the issue with accessing the DetectorData objects via python. I will look into this...
